### PR TITLE
740 Publishing service pushes policy to minio

### PR DESCRIPTION
**Pull request by AleF83**

### CODE REVIEW

File: CI/docker-compose.yml [hunk 1]
```diff
@@ -96,14 +90,14 @@ services:
...
-      - TWEEKGATEWAY_SECURITY_POLICYSTORAGE_MINIOBUCKETNAME=tweek-casbin-policies
+      - TWEEKGATEWAY_SECURITY_POLICYSTORAGE_MINIOBUCKETNAME=tweek-bucket
       - TWEEKGATEWAY_SECURITY_POLICYSTORAGE_MINIOACCESSKEY=AKIAIOSFODNN7EXAMPLE
       - TWEEKGATEWAY_SECURITY_POLICYSTORAGE_MINIOSECRETKEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
       - TWEEKGATEWAY_SECURITY_POLICYSTORAGE_MINIOSECURE=false
       - TWEEKGATEWAY_SECURITY_POLICYSTORAGE_MINIOPOLICYOBJECTNAME=policy.csv
       - TWEEKGATEWAY_SECURITY_POLICYSTORAGE_NATSENDPOINT=nats://nats:4222
-      - TWEEKGATEWAY_SECURITY_POLICYSTORAGE_NATSSUBJECT=casbin-policy-subject
-      - TWEEKGATEWAY_SECURITY_TWEEKSECRETKEYPATH=/gitkeys/tweekgit
+      - TWEEKGATEWAY_SECURITY_POLICYSTORAGE_NATSSUBJECT=version
```
Comment (Yshayy) [inline CI/docker-compose.yml]: it should be hardcoded.

Comment (Yshayy) [reply CI/docker-compose.yml]: Yet, we should have support for nats prefix so we can run on external NAT or run several Tweek on the same NAT

File: CI/docker-compose.yml [hunk 2]
```diff
@@ -23,6 +20,7 @@ services:
       - MINIO__Endpoint=minio:9000
       - MINIO__AccessKey=AKIAIOSFODNN7EXAMPLE
       - MINIO__SecretKey=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
+      - MINIO__Bucket=tweek-bucket
```
Comment (Yshayy) [inline CI/docker-compose.yml]: We can drop "bucket" from "tweek-bucket"

File: e2e/integration/spec/authoring-api/policy.spec.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+describe('authoring api policy', () => {
+  let clients;
+  before(async () => {
+    clients = await initClients();
+  });
+
+  after(async () => {
+    const policy = 'p, test@tweek.com, /api/v2/values/*, *';
+
+    await clients.authoring
```
Comment (Yshayy) [inline e2e/integration/spec/authoring-api/policy.spec.js]: Should require "Admin" by default, like apps endpoint (createApp)

Comment (Yshayy) [reply e2e/integration/spec/authoring-api/policy.spec.js]: Should create policy that doesn't affect other tests.

Comment (Yshayy) [reply e2e/integration/spec/authoring-api/policy.spec.js]: Will also be unnecessary once #775 is merged

File: e2e/integration/utils/clients.js [hunk 4]
```diff
@@ -33,14 +31,24 @@ nconf
...
     GATEWAY_URL: 'http://localhost:4099',
     GIT_PRIVATE_KEY_PATH: '../../deployments/dev/ssh/tweekgit',
+
+    MINIO_HOST: 'localhost',
+    MINIO_PORT: '4007',
+    MINIO_BUCKET: 'tweek-bucket',
   });
 
-module.exports.init = async function() {
-  const inlineKey = nconf.get('GIT_PRIVATE_KEY_INLINE');
```
Comment (Yshayy) [inline e2e/integration/utils/clients.js]: check why we switched implementation

File: e2e/integration/utils/minio.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  const mc = new minio.Client({
+    endPoint: nconf.get('MINIO_HOST'),
+    port: Number(nconf.get('MINIO_PORT')),
+    accessKey: getEnv('MINIO_ACCESS_KEY'),
+    secretKey: getEnv('MINIO_SECRET_KEY'),
+    secure: false,
+  });
+
+  return new Promise((resolve, reject) =>
+    mc.getObject(nconf.get('MINIO_BUCKET'), objectName, (err, stream) => {
```
Comment (Yshayy) [inline e2e/integration/utils/minio.js]: check promisify or bluebird stream utils

File: services/authoring/src/routes/policies.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+@AutoWired
+@Path('/policies')
+export class PolicyController {
+  @Context
+  context: ServiceContext;
+
+  @Inject
+  policyRepository: PolicyRepository;
+
+  @Authorize({ permission: PERMISSIONS.ADMIN })
```
Comment (Yshayy) [inline services/authoring/src/routes/policies.ts]: Admin is already specified, ignore previous comment regrading that

File: services/publishing/Tweek.Publishing.Service/Sync/StorageSynchronizer.cs [hunk 7]
```diff
@@ -64,31 +59,10 @@ public StorageSynchronizer(IObjectStorage storageClient, ShellHelper.ShellExecut
...
             };
 
-            var (p, exited) = _shellExecutor("git", $"archive --format=zip {commitId}");
-            using (var ms = new MemoryStream())
+            foreach (var uploader in Uploaders)
             {
-                await p.StandardOutput.BaseStream.CopyToAsync(ms);
-                ms.Position = 0;
-                await exited;
+                await uploader.Upload(commitId);
```
Comment (Yshayy) [inline services/publishing/Tweek.Publishing.Service/Sync/StorageSynchronizer.cs]: We should consider paralleling this in the future.

File: services/publishing/Tweek.Publishing.Service/Sync/Uploaders/RulesUploader.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                using (var zip = new ZipArchive(ms, ZipArchiveMode.Read, false))
+                {
+                    var files = zip.Entries.Select(x => x.FullName).ToList();
+                    var bundle = _packer.Pack(files, GetZipReader(zip));
+                    await _client.PutJSON(commitId, bundle);
+                }
+            }            
+        }
+
+        private static Func<string, string> GetZipReader(ZipArchive zip)
```
Comment (Yshayy) [inline services/publishing/Tweek.Publishing.Service/Sync/Uploaders/RulesUploader.cs]: use arrow

File: services/publishing/Tweek.Publishing.Service/Sync/StorageSynchronizer.cs (hunk 7 above)
Comment (AleF83) [reply services/publishing/Tweek.Publishing.Service/Sync/StorageSynchronizer.cs]: issue added #776

File: e2e/integration/utils/minio.js (hunk 5 above)
Comment (AleF83) [reply e2e/integration/utils/minio.js]: I checked it.. It doesn't exist

**Pull request merged**